### DISCUSSION TITLE
[Feature Request] Build args support in Docker container actions

### DISCUSSION THREAD

**Discussion by yoannchaudet** [community/community#25241] [Actions]
<p>Unless I missed it, it does not seem possible today to specify Docker build arguments (i.e. arguments passed when building an action’s container.</p>
<p>Being able to specify a “<a href="https://docs.docker.com/engine/reference/commandline/build/" rel="nofollow noopener">build-args</a>” section in an <a href="https://help.github.com/en/actions/automating-your-workflow-with-github-actions/creating-a-docker-container-action#create-an-action-metadata-file" rel="nofollow noopener">action metadata file</a> would be nice.</p>
<p>Potential scenario: specify the tag of base image with a variable (whose value would be defined in the action metadata file and potentially controlled by inputs or environment variables).</p>

**Comment by MChevy422** [comment]
<p>Hello <span><a class="mention" href="/yoannchaudet">@yoannchaudet</a></span>, <br><br>Thank you for your feedback! We’re always working to improve GitHub and the GitHub Community Forum, and we consider every suggestion we receive. I’ve logged your feature request in our internal feature request list. Though I can’t guarantee anything or share a timeline for this, I can tell you that it’s been shared with the appropriate teams for consideration.<br><br>Once again, thank you for your input!<br><br>Greatly appreciated<br><br>MChevy422</p>

**Reply by collinkleest** [reply]
@MChevy422 has there been any updates on this, I know this is old. A lot of developers would use this.

**Comment by BobbyMcWho** [comment]
<p>Yes, PLEASE! This is a blocker for something I want to do in an action, and it’s tough to  find a non-hacky feeling solution.</p>

**Comment by dgw** [comment]
<p>Hi <a class="mention" href="/mchevy422">@mchevy422</a>,</p>
<p>Please tally an extra vote for this feature on my behalf!</p>
<p>For further context, I came across this thread while searching for a way to let the action user select what version of the base image to use for a Docker action. That is, it would default to the base image’s “latest” tag, but something like this step in a job would select a different version tag and build the action using that instead:</p>
<pre><code>- name: Run my-action with custom base image version
        uses: dgw/my-action@v1
        with:
          base: 1.2.8
</code></pre>
<p>And base could be passed back to the Dockerfile build via (for example):</p>
<pre><code>inputs:
  base:
    description: 'Base Docker image version to use'
    required: true
    default: latest
</code></pre>
<p>(Could also add a new field to with, but to me it makes more sense that the version would be an input.)</p>
<p>Thanks!</p>

**Comment by jdolitsky** [comment]
<p>Specifically, I’d like to see the build arg to have access to the git commit of the source repo, for example</p>
<pre><code>runs:
  using: 'docker'
  image: 'Dockerfile'
  build-args:
    - name: VERSION
      value: ${{ thisrepo.gitcommit }}
</code></pre>
<p>Especially when building off of somebody’s “<span class="mention">@master</span>” version of the action, would be nice to know the git commit that built it.</p>

**Comment by JavierZolotarchuk** [comment]
<p>A work around until the feature is implemented is that we use the docker image that creates the action as handrails to create the image that we really want with all the arguments we need.<br>
Basically what we have to do is create a docker image in the entrypoint.sh and run it.</p>
<p><em>entrypoint.sh example:</em></p>
<pre><code>#!/bin/sh -l

ALPINE_VERSION=$1

cd docker-action

echo "creating docker image with alpine version: $ALPINE_VERSION"

# here we can make the construction of the image as customizable as we need
# and if we need parameterizable values ​​it is a matter of sending them as inputs

docker build -t docker-action --build-arg alpine_version="$ALPINE_VERSION" . &amp;&amp; docker run docker-action
</code></pre>
<p>For the entrypoint.sh to work correctly it is necessary that the necessary dependencies are installed on the main image to be able to use docker ( <strong>I recommend that the main image be an alpine so that it is as light as possible</strong> ).</p>
<p><strong>NOTE:</strong> By main image I mean the image resulting from built the DockerFile referenced in the action.yml file.</p>
<p>Here is a simple example of how to do it.</p>
<p><a href="https://github.com/JavierZolotarchuk/parameterizable-docker-action-example" rel="nofollow noopener">https://github.com/JavierZolotarchuk/parameterizable-docker-action-example</a></p>

**Comment by JavierZolotarchuk** [comment, marked as answer]
<p>A work around until the feature is implemented is that we use the docker image that creates the action as handrails to create the image that we really want with all the arguments we need.<br>
Basically what we have to do is create a docker image in the entrypoint.sh and run it.</p>
<p><em>entrypoint.sh example:</em></p>
<pre><code>#!/bin/sh -l

ALPINE_VERSION=$1

cd docker-action
echo "creating docker image with alpine version: $ALPINE_VERSION"

# here we can make the construction of the image as customizable as we need
# and if we need parameterizable values ​​it is a matter of sending them as inputs
docker build -t docker-action --build-arg alpine_version="$ALPINE_VERSION" . &amp;&amp; docker run docker-action
</code></pre>
<p>For the entrypoint.sh to work correctly it is necessary that the necessary dependencies are installed on the main image to be able to use docker ( <strong>I recommend that the main image be an alpine so that it is as light as possible</strong> ).</p>
<p><strong>NOTE:</strong> By main image I mean the image resulting from built the DockerFile referenced in the action.yml file.</p>
<p>Here is a simple example of how to do it.</p>
<p><a href="https://github.com/JavierZolotarchuk/parameterizable-docker-action-example" rel="nofollow noopener">https://github.com/JavierZolotarchuk/parameterizable-docker-action-example</a></p>

**Reply by fcano-ut** [reply]
This seems like a waste of resources. You need to install docker inside a docker container, just to be able to pass build args.
Also, ideally I'd expect GitHub to be able to "cache" the built image somehow instead of building from scratch every time the action is used. But in this approach, the image needs to be built every time.

Not trying to undermine your solution (I'm quite grateful that there's a workaround actually) but just saying that this is just a "patch" until we have feature support.

**Reply by adamjakab** [reply]
As @JavierZolotarchuk says. this is a workaround and not a solution / answer. 
Marking this as the answer 4 years ago achieved only one thing... that nobody is looking at this because this is a thread with an answer. Pls remove the "answered" tag.

**Comment by mmailhos** [comment]
<p>This is needed.</p>
<p>Use case: run some code analysis (or any language specific tools) with the same Docker image version as the language version used in the repo (such as running <em>FROM golang:1.13.2</em> in a repo using Go 1.13.2). The only way to do this is to pass the docker image tag from the Action input to the <em>build-arg</em>.</p>

**Comment by w9jds** [comment]
<p>I would like to see this as well. In my actions <code>firebase-actions</code> a common request is to allow you to run the emulator, as select the node version (as functions requires a very specific version). For the emulator I would like to avoid including the JDK unless absolutely needed as it will increase build times quite a bit. I need the build-args to allow me to provide them the ability to flag for emulation (install the JDK) and set the version of node they would like the container to use.</p>

**Comment by aaronzs** [comment]
<p>I’d like to see this feature as well. My current solution is using <code>composite</code>.</p>
<pre><code class="lang-yaml">runs:
  using: 'composite'
  steps:
    - id: build
      run: docker build . &lt;build-args&gt; -t image:tag
      shell: bash
    - id: run
      run: docker run --rm -v $(pwd):/github/workspace --workdir /github/workspace image:tag
      shell: bash
</code></pre>

**Reply by nemith** [reply]
`docker run` is unfortunate if you rely on on any environment variables as you have to whitelist everone of them with `-e` or do some hackery with `.env` files.

**Comment by zmingxie** [comment]
<p>Thanks for the suggestion from <a class="mention" href="/javierzolotarchuk">@javierzolotarchuk</a></p>
<p>I managed to get my GitHub action updated with an optional build version argument using the suggestion: <a href="https://github.com/zmingxie/packer-ami-builder" rel="noopener nofollow ugc">https://github.com/zmingxie/packer-ami-builder</a></p>
<p>However, this approach causes one side-effect which we won’t be able to live tail the build log within the GitHub Action web console. Logs will only be accessible after the build is completed.</p>
<p>Really hope GitHub will add this feature soon.</p>

**Comment by bmacnaughton** [comment]
<p>thanks for this nice workaround. it solved my issue though a github actions solution would not require the intermediate container (as you know) and would be much cleaner. things people should be aware of:</p>
<ul>
<li>it will not have the GITHUB_WORKSPACE or any of the other directories mapped</li>
<li>final container will not have all the GITHUB environment variables</li>
<li>you’ll have to manually pass arguments from one entrypoint.sh to the other</li>
</ul>
<p>you can manually map the various directories and pass env vars.</p>

**Comment by bmacnaughton** [comment]
<p>i originally tried to solve this with a slightly different approach, using</p>
<pre><code class="lang-auto">    runs:
        using: 'docker'
        image: '${{matrix.os-version}}.Dockerfile'
</code></pre>
<p>but that syntax is not accepted. it’s a bit of a mystery to me just where interpolation is allowed and where it is not allowed. ultimately, i ended up using <a href="https://github.community/u/javierzolotarchuk">javierzolotarchuk</a>’s solution.</p>

**Comment by mvarrieur** [comment]
<p>Also ran into this today, the workarounds don’t work for my use case so we are stuck unfortunately.</p>

**Comment by clarencedesmarais** [comment]
<p>Ran into this issue today. Was wondering how to actually pass ARG in my docker container. Any updates on this ?</p>

**Comment by tmtenbrink** [comment]
<p>This would be really helpful, especially if it also allows passing --cache-to and --cache-from arguments.</p>

**Comment by set-soft** [comment]
<p>I tried manually mapping directories and it doesn’t work.<br>
I can’t pass data from the intermediate container to the second one.<br>
Mapped volumes aren’t visible and any data I write in new volumes just disapear as soon as the second contianer finishes.<br>
Any idea?</p>

**Comment by nishit130** [comment]
Hey, any update on the feature request? Would love to see it. I have the exact same use case as mentioned by op. I want to install a dependency inside Dockerfile which needs to have dynamic version (i.e version to be picked up by env).

**Comment by ccfontes** [comment]
This looks to me more like a bug than a feature request.

**Comment by madhums** [comment]
Any update on this? This is much needed!

**Comment by gabrielrufino** [comment]
Up

**Comment by zachspar** [comment]
Would love to see this feature.

**Comment by MarwaOuledElHadjAli** [comment]
Any updated on this feature request please ?
The issue was marked as answered but it wasn't the case

**Comment by itsMarcoSolis** [comment]
Hi @yoannchaudet could you please mark this as unanswered to see if we can get this moving towards a proper solution?

**Comment by madhums** [comment]
Can we please please please find a better solution for this?
